internal/nbd/nbdserver.go

```diff
@@ -0,0 +1,315 @@
+package nbd
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/coreos/pkg/capnslog"
+)
+
+var clog = capnslog.NewPackageLogger("github.com/coreos/torus", "nbd")
+
+const (
+	nbdMagic       uint64 = 0x4e42444d41474943
+	nbdOpts        uint64 = 0x49484156454F5054
+	nbdRequestSize        = (4 + 4 + 8 + 8 + 4)
+
+	nbdFlagFixedNewStyle = 0x1
+
+	// options
+	nbdOptExportName = 0x1
+	nbdOptAbort      = 0x2
+	nbdOptList       = 0x3
+
+	// option replies
+	nbdRepAck      = 0x1
+	nbdRepServer   = 0x2
+	nbdRepErrUnsup = 0x80000001
+
+	ioKeepAlive = 10 * time.Second
+)
+
+type DeviceFinder interface {
+	FindDevice(name string) (Device, error)
+	ListDevices() ([]string, error)
+}
+
+type NBDServer struct {
+	l      *net.TCPListener
+	finder DeviceFinder
+}
+
+func NewNBDServer(addr string, finder DeviceFinder) (*NBDServer, error) {
+	tcpaddr, err := net.ResolveTCPAddr("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	ln, err := net.ListenTCP("tcp", tcpaddr)
+	if err != nil {
+		return nil, err
+	}
+
+	ns := &NBDServer{
+		l:      ln,
+		finder: finder,
+	}
+
+	return ns, nil
+}
+
+func (s *NBDServer) Serve() error {
+	for {
+		c, err := s.l.AcceptTCP()
+		if err != nil {
+			return err
+		}
+
+		c.SetKeepAlive(true)
+		c.SetKeepAlivePeriod(ioKeepAlive)
+
+		conn := &NBDConn{
+			c:      c,
+			finder: s.finder,
+			export: "<none>",
+		}
+
+		go func() {
+			if err := conn.handler(); err != nil {
+				conn.errorf("handler error: %v", err)
+			}
+			conn.Close()
+		}()
+	}
+}
+
+func (s *NBDServer) Close() error {
+	return s.l.Close()
+}
+
+type option struct {
+	opt  uint32
+	data []byte
+}
+
+type NBDConn struct {
+	c      *net.TCPConn
```

Comment:
i know that. we already set keep alive before putting this into ndbconn. not a big deal anyway. it is up to you. probably we will only have tcp ndb conn in foreseeable future...